lgc/patch/NggPrimShader.cpp

```diff
@@ -5998,6 +6010,76 @@ Value *NggPrimShader::doDppUpdate(Value *oldValue, Value *srcValue, unsigned dpp
                                      m_builder->getInt32(bankMask), m_builder->getInt1(boundCtrl)});
 }
 
+// =====================================================================================================================
+// Fetches the position data for the specified vertex ID.
+//
+// @param vertexId : Vertex thread ID in sub-group.
+Value *NggPrimShader::fetchVertexPositionData(Value *vertexId) {
+  Value *posData = nullptr;
+
+  if (m_hasGs) {
```

Comment:
I want to preface this comment by saying that I don't want to nitpick too much and won't block this PR in the current state if you are happy with it.

I think the current revision of this function is not a strict improvement over the previous one. Instead of two returns for each execution path, now we have one with extra data-flow. When reading or debugging code like this, one of the basic question for me is 'what values does this function return'. To answer this in the current form, you have to look at the single `return` and then track back all possible assignments to `posData`, and have to reason about control- and data-flow at the same time. This is not a big deal at all for this function because it's quite simple, but can lead complicated spaghetti as the code grows. I tried to address this exact issue in one of my recent PR's: #596. I think that the code would be much more maintainable if each execution path had it's own return statement and there was no infeasible control flow path with no return at all, like in the first revision of this PR.

If you structured the code like this:
```
Value *Class::foo() {
   if (...) {
    // ES-GS
    // ...
    return bar(...);
  }
  
  // ES-only
  return baz(...);
}
```
the code would look much easier to reason about and debug for me. If I wanted to know which execution path was taken, I could just set breakpoints on the two returns, and I wouldn't have to worry about the function not returning in some other execution path. Again, this is not a big deal at all in this particular case, but this is a more general observation that I made when reading a lot of existing code that declares a `Result result = Result::Success` and the beginning and then keeps branching on `results` -- in my view, a lot of code like this would benefit from a similar refactoring as in #596.

Reply:
It is no nitpicking. I think it is OK to discuss the coding philosophy. What you have seen in original codes is caused by PAL style. The style requires only one return is allowed for each function. The reason is to have only one common exit in a function for debugging. You don't have to place many breakpoints on all returns to catch the run-time calculated return values. So the logic is often something like this:
```
func()
{
    result = Initial Value;
    if (result == Success)
    {
        // Do something
        result = Y;
    }
    if (result == Success)
    {
        // Do something
        result = Z;
    }
    ...
    return result;
}
```
Admittedly, different developers have different strategies. I will change the codes as you requested.

Reply:
I finally adopt your suggestion in the initial review. Take ES-only path as a special one
```
func() {
  if (!m_hasGs) {
    // ES-only
  }
  // ES-GS
}

```